### DISCUSSION TITLE
Upcoming changes to block.number, block.timestamp & blockHash

### DISCUSSION THREAD

**Discussion by uF4No** [zkSync-Community-Hub/zksync-developers#32] [Announcements]
**UPDATE**
[All details and upgrade plan have been shared in this new thread](https://github.com/zkSync-Community-Hub/zkync-developers/discussions/87).

**Original message below**

We’re making important changes to how `block.number`, `block.timestamp`, and `blockHash` behave on zkSync Era. Here is what you need to know: 

### 🔀 Scope of Changes

`block.number`, `block.timestamp` and `blockHash` currently return the number, timestamp and hash of the L1 batch. **After these changes they will return the number, timestamp and hash of the L2 block**. We will introduce a new set of methods for accessing values in the L1 batch.
 

### ⏱️ Background & Motivation

Many applications require a higher fidelity of time that can be achieved by referencing L1 batches. With these changes, contracts will be able to measure time at the level of *L2 blocks* (produced ≈ every 1 - 2 seconds!). This information is [already available on the API](https://era.zksync.io/docs/reference/concepts/transactions/blocks.html#retrieving-block-and-batch-numbers) but with the upcoming changes, developers will be able to access it directly in the smart contracts. For consistency, `block.number` and `blockHash` should also correspond to the L2 block. 

### 🚨 Impact on Projects

**block.number**

The number of L2 blocks far exceeds the number of L1 batches (each batch contains a set of blocks). This means that immediately after the changes have been applied there will be one-time *jump* in the value returned by `block.number`. For example: if the changes were applied in [this batch](https://explorer.zksync.io/batch/74126) — it could change from [74126](https://explorer.zksync.io/batch/74126) → [652929](https://explorer.zksync.io/block/6529295). After this jump, `block.number` will only return values from the L2 block.

**block.timestamp**

The timestamp of L1 batches is roughly equivalent to the timestamp for L2 batches. So while `block.timestamp` will also *jump* from batches to L2 blocks, the change will be less significant. After this point, `block.timestamp` will be updated at the rate of L2 block production (≈ every 1 -2 seconds).

**blockHash**

This upgrade will change the structure of `blockHash`. The hash of the L1 batch is the root of the merkle tree. The hash of the *L2 block*, however, has a different composition (a mix of timestamp block number, and potentially additional data about the L2 block). 

### 🗓️ Upgrade Timeline

Instead of providing a specific date right away, we would like to involve the developer community in gathering feedback first. However, as a rough estimate, we anticipate implementing these changes sometime between July and August.

### 💬 Feedback

If you have any feedback on these changes we would love to hear from you!

**Comment by deanpress** [comment]
Thank you for working on this. I (as I'm sure many developers) are happy to hear timestamps and blockhashes will be more in key with what we expect.

For current development I had to build an [L2 block data oracle](https://gist.github.com/deanpress/e139c44d43e6ee34fbbe125a47611459) just so we could get our contracts working with zkSync. I would much prefer to use on-chain data for this rather than having to build completely separate versions of each contract I'm working on just to get it working with zkSync.

Regarding blockhash, can we expect it to offer similar PRNG to L1 blockhash? I noticed you said "potentially additional data about the L2 block". Preferably this is as much data of the L2 block as possible.

Also, what will the history limit be for blockhashes? On L1 it's 256 but some L2 networks have it infinitely available.

**Reply by uF4No** [reply]
The blockhash will include number, timestamp and hashes of the transactions. Unlike Ethereum, it will not include the state root. We would not encourage the use of blockhash as a source for randomness. You can use it but keep in mind it might not be secure.

> what will the history limit be for blockhashes? 

It will be 256

**Reply by senol1985** [reply]
🚨🚨🚨zz coin SCAMM🚨🚨🚨

**Comment by KonScanner** [comment]
Suppose I wanted to say to initialize a `contract_address` and call `balanceOf` for a specific block. Will it be that before a certain point, I need to use the L1 block number and then use the L2 block number?

**Reply by githubdoramon** [reply]
@KonScanner in this case you mean calling it from the API, not on solidity right? If this is the case, the API already works with L2 number and timestamp, so nothing changes there.

**Reply by senol1985** [reply]
🚨🚨🚨zz coin SCAMM🚨🚨🚨

**Comment by nishuzumi** [comment]
Hello sir, this block number modification is nothing but a heavy blow to some DEFI applications, can I know the exact timing otherwise some protocols that distribute rewards as per block number may have huge problems.

**Reply by senol1985** [reply]
🚨🚨🚨zz coin SCAMM🚨🚨🚨

**Reply by uF4No** [reply]
@nishuzumi We wanted to create this thread to gather feedback from projects (thanks!) and understand potential issues and how these changes might affect different projects across the ecosystem.

We understand these changes might impact some projects so we won't make the update without a proper notice period.

**Comment by senol1985** [comment]
![Screenshot_20230715-140703_MEXC](https://github.com/zkSync-Community-Hub/zkync-developers/assets/90096660/f7d55193-343f-4c93-ba8f-5146a883f82d)
🚨🚨🚨 zz coin SCAMM 🚨🚨🚨

**Comment by uF4No** [comment]
We've just shared [all the details of the upgrade plan in this new thread](https://github.com/zkSync-Community-Hub/zkync-developers/discussions/87)

**Comment by 333david** [comment]
Zksync is the best layer 2 product

**Reply by jmwhangjin** [reply]
This is the best

**Comment by mnuww** [comment]
keep going

**Comment by edevaldonunez** [comment]
This update regarding block.number, block.timestamp, and blockHash on zkSync Era is a significant improvement for developers, providing higher fidelity time measurement and aligning block-related values with L2 blocks. The transition plan outlined with anticipated changes between July and August allows for community feedback, ensuring smooth integration and minimal disruption to projects. Exciting advancements ahead!

**Comment by jecgaegacharsmi** [comment]
Important updates to block.number, block.timestamp, and blockHash behavior on zkSync Era. Changes will align these values with L2 blocks instead of L1 batches. Feedback from the developer community is welcomed before implementation, expected between July and August.

**Comment by hunastly** [comment]
Important update: Changes to block.number, block.timestamp, and blockHash behavior on zkSync Era are coming. Get ready for improved time fidelity and consistency! Learn more and share your feedback: [link to thread] #zkSync #Upgrade #Web3
